solutions/cf-1017G.mdx

```diff
@@ -0,0 +1,218 @@
+
+Note that for query operations, we need to find the maximum suffix sum for each path. This can be done by maintaing the maximum suffix sum as well as the actual sum in each segment tree node. 
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(n + q\log^2n)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAXN = 100001;
+
+using ll = long long
+using pll = pair<ll, ll>;
+#define FF first
+#define SS second
+
+int N, Q;
+vector<int> G[MAXN];
+
+// Segment Tree:
+namespace sgt {
+	pll T[MAXN * 4]; // {value, sum}
+	bool L[MAXN * 4]; // set tag
+
+	pll merge(pll l, pll r) {
+		if (l.SS == -LLONG_MAX) {
+			return r;
+		}
+		if (r.SS == -LLONG_MAX) {
+			return l;
+		}
+		return {max(l.FF + r.SS, r.FF), l.SS + r.SS};
+	}
+	
+	void build(int t = 1, int tl = 0, int tr = N - 1) {
+		if (tl == tr) {
+			T[t].FF = T[t].SS = -1;
+			return;
+		}
+
+		int tm = (tl + tr) >> 1;
+		build(t << 1, tl, tm);
+		build(t << 1 | 1, tm + 1, tr);
+		T[t] = merge(T[t << 1], T[t << 1 | 1]);
+	}
+
+	void pushdown(int t, int tl, int tr) {
+		if (!L[t]) {
+			return;
+		}
+		int tm = (tl + tr) >> 1;
+		T[t << 1] = {-1, -(tm - tl + 1)};
+		T[t << 1 | 1] = {-1, -(tr - tm)};
+		L[t << 1] = L[t << 1 | 1] = 1;
+		L[t] = 0;
+	}
+
+	void update_range(int l, int r, int t = 1, int tl = 0, int tr = N - 1) {
+		if (r < tl || tr < l) {
+			return;
+		}
+		if (l <= tl && tr <= r) {
+			T[t] = {-1, -(tr - tl + 1)};
+			L[t] = 1;
+			return;
+		}
+		pushdown(t, tl, tr);
+
+		int tm = (tl + tr) >> 1;
+		update_range(l, r, t << 1, tl, tm);
+		update_range(l, r, t << 1 | 1, tm + 1, tr);
+		T[t] = merge(T[t << 1], T[t << 1 | 1]);
+	}
+
+	void update_point(int i, int v, int t = 1, int tl = 0, int tr = N - 1) {
+		if (tl == tr) {
+			T[t].FF += v;
+			T[t].SS += v;
+			return;
+		}
+		pushdown(t, tl, tr);
+
+		int tm = (tl + tr) >> 1;
+		if (i <= tm) {
+			update_point(i, v, t << 1, tl, tm);
+		} else {
+			update_point(i, v, t << 1 | 1, tm + 1, tr);
+		}
+		T[t] = merge(T[t << 1], T[t << 1 | 1]);
+	}
+
+	pll query(int l, int r, int t = 1, int tl = 0, int tr = N - 1) {
+		if (r < tl || tr < l) {
+			return {-LLONG_MAX, -LLONG_MAX};
+		}
+		if (l <= tl && tr <= r) {
+			return T[t];
+		}
+		pushdown(t, tl, tr);
+
+		int tm = (tl + tr) >> 1;
+		return merge(query(l, r, t << 1, tl, tm), query(l, r, t << 1 | 1, tm + 1, tr));
+	}
+}
+
+	
+// Heavy-Light Decomposition:
```

Comment:
ppl who read this should know about hld right?

Reply:
seeing as it's in the HLD section, yes